rust/main/chains/hyperlane-sovereign/src/merkle_tree_hook.rs

```diff
@@ -1,10 +1,129 @@
-use crate::{ConnectionConf, Signer, SovereignProvider};
+use crate::{
+    indexer::SovIndexer,
+    rest_client::{to_bech32, SovereignRestClient, TxEvent},
+    ConnectionConf, Signer, SovereignProvider,
+};
 use async_trait::async_trait;
+use core::ops::RangeInclusive;
 use hyperlane_core::{
-    accumulator::incremental::IncrementalMerkle, ChainResult, Checkpoint, ContractLocator,
-    HyperlaneChain, HyperlaneContract, HyperlaneDomain, HyperlaneProvider, MerkleTreeHook, H256,
+    accumulator::incremental::IncrementalMerkle, ChainCommunicationError, ChainResult, Checkpoint,
+    ContractLocator, HyperlaneChain, HyperlaneContract, HyperlaneDomain, HyperlaneProvider,
+    Indexed, Indexer, LogMeta, MerkleTreeHook, MerkleTreeInsertion, SequenceAwareIndexer, H256,
+    H512,
 };
-use std::num::NonZeroU64;
+use serde::Deserialize;
+
+use std::{num::NonZeroU64, str::FromStr};
+
+/// Struct that retrieves event data for a Cosmos Mailbox contract
+#[derive(Debug, Clone)]
+pub struct SovereignMerkleTreeHookIndexer {
+    provider: Box<SovereignProvider>,
+    bech32_address: String,
+}
+
+impl SovereignMerkleTreeHookIndexer {
+    pub async fn new(
+        conf: ConnectionConf,
+        locator: ContractLocator<'_>,
+        _signer: Option<Signer>,
+    ) -> ChainResult<Self> {
+        let provider = SovereignProvider::new(locator.domain.clone(), &conf, None).await;
+        Ok(SovereignMerkleTreeHookIndexer {
+            provider: Box::new(provider),
+            bech32_address: to_bech32(locator.address)?,
+        })
+    }
+}
+
+#[async_trait]
+impl crate::indexer::SovIndexer<MerkleTreeInsertion> for SovereignMerkleTreeHookIndexer {
+    const EVENT_KEY: &'static str = "Merkle/InsertedIntoTree";
+
+    fn client(&self) -> &SovereignRestClient {
+        self.provider.client()
+    }
+
+    async fn sequence_at_slot(&self, slot: u32) -> ChainResult<Option<u32>> {
+        let sequence = self
+            .client()
+            .tree(&self.bech32_address, NonZeroU64::new(u64::from(slot)))
+            .await?;
+
+        match u32::try_from(sequence.count) {
+            Ok(x) => Ok(Some(x)),
+            Err(e) => Err(ChainCommunicationError::CustomError(format!(
+                "Tree count error: {e:?}"
+            ))),
+        }
+    }
+    fn decode_event(&self, event: &TxEvent) -> ChainResult<MerkleTreeInsertion> {
+        let parsed_event: InsertedIntoTreeEvent = serde_json::from_value(event.value.clone())?;
+
+        let merkle_insertion = MerkleTreeInsertion::new(
+            parsed_event
+                .clone()
+                .inserted_into_tree
+                .and_then(|d| d.index)
+                .ok_or_else(|| {
+                    ChainCommunicationError::CustomError(String::from(
+                        "parsed_event contained None",
+                    ))
+                })?,
```

Comment:
Can we avoid cloning here?

```suggestion
            parsed_event
                .inserted_into_tree
                .as_ref()
                .and_then(|d| d.index)
                .ok_or_else(|| {
                    ChainCommunicationError::CustomError(String::from(
                        "parsed_event contained None",
                    ))
                })?,
```